enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/machines/id/IdReusabilityCondition.java

```diff
@@ -0,0 +1,103 @@
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.causalclustering.core.state.machines.id;
+
+import java.util.function.BooleanSupplier;
+
+import org.neo4j.causalclustering.core.consensus.RaftMachine;
+import org.neo4j.causalclustering.identity.MemberId;
+import org.neo4j.kernel.impl.util.Listener;
+
+/**
+ * Determines whether it is safe to reuse freed ids, based on current leader and tracking its own transactions.
+ * This should guarantee that a single freed id only ends up on a single core.
+ */
+public class IdReusabilityCondition implements BooleanSupplier, Listener<MemberId>
+{
+    private static final BooleanSupplier ALWAYS_FALSE = () -> false;
+
+    private CommandIndexTracker commandIndexTracker;
+    private final RaftMachine raftMachine;
+    private final MemberId myself;
+
+    private volatile BooleanSupplier currentSupplier = ALWAYS_FALSE;
+
+    public IdReusabilityCondition( CommandIndexTracker commandIndexTracker, RaftMachine raftMachine, MemberId myself )
+    {
+        this.commandIndexTracker = commandIndexTracker;
+        this.raftMachine = raftMachine;
+        this.myself = myself;
+        raftMachine.registerListener( this );
+    }
+
+    @Override
+    public boolean getAsBoolean()
+    {
+        return currentSupplier.getAsBoolean();
+    }
+
+    @Override
+    public void receive( MemberId newLeader )
+    {
+        if ( myself.equals( newLeader ) )
+        {
+            // We just became leader
+            currentSupplier = new LeaderIdReusabilityCondition( commandIndexTracker, raftMachine );
+        }
+        else
+        {
+            // We are not the leader
+            currentSupplier = ALWAYS_FALSE;
+        }
+    }
+
+    private static class LeaderIdReusabilityCondition implements BooleanSupplier
+    {
+        private final CommandIndexTracker commandIndexTracker;
+        private final long commandIdWhenBecameLeader;
+
+        private volatile boolean oldTransactionsApplied;
+
+        LeaderIdReusabilityCondition( CommandIndexTracker commandIndexTracker, RaftMachine raftMachine )
+        {
+            this.commandIndexTracker = commandIndexTracker;
+
+            // Get highest command id seen
+            this.commandIdWhenBecameLeader = raftMachine.state().lastLogIndexBeforeWeBecameLeader();
+        }
+
+        @Override
+        public boolean getAsBoolean()
+        {
+            if ( oldTransactionsApplied )
+            {
+                return true;
+            }
+
+            boolean hasAppliedOldTransactions = commandIndexTracker.getAppliedCommandIndex() > commandIdWhenBecameLeader;
+            if ( hasAppliedOldTransactions )
+            {
+                oldTransactionsApplied = true;
+            }
+
+            return hasAppliedOldTransactions;
```

Comment:
This code can be simplified. Would be nice with a comment describing the deeper meaning behind this logic.